### PR TITLE
[PUBDEV-7859] add ability to resume grid after cluster crash

### PR DESCRIPTION
**Pull request by honzasterba**
https://h2oai.atlassian.net/browse/PUBDEV-7859

### CODE REVIEW

File: h2o-core/src/main/java/hex/Model.java
Comment (Pscheidl) [inline h2o-core/src/main/java/hex/Model.java:412]: Knit/opinion based:

- The collection of two basic frames (validation and train) could be part of the method's contract in JavaDoc, **or** there could be yet another method `collectCommonFrames`, which would make it obvious for overriders what is done. and could selectively pick functionaly to override. Calling `super` is fine to me as well.

- MIssingJavaDoc.

File: h2o-core/src/main/java/hex/Model.java
Comment (Pscheidl) [inline h2o-core/src/main/java/hex/Model.java:413]: Should the type of the frame be an enum ? Possibly algo-specific ? Java doesn't support enum inheritance, could be done with constants in `FrameType` class or something like that. Just thinking out loud.

File: h2o-core/src/main/java/water/fvec/persist/PersistUtils.java
Comment (Pscheidl) [inline h2o-core/src/main/java/water/fvec/persist/PersistUtils.java:22]:
```suggestion
    public static void write(URI uri, Writer w) throws RuntimeException {
```

File: h2o-core/src/main/java/water/fvec/persist/PersistUtils.java
Comment (Pscheidl) [inline h2o-core/src/main/java/water/fvec/persist/PersistUtils.java:34]:
```suggestion
    public static void write(URI uri, Writer w) throws RuntimeException {
```

File: h2o-core/src/main/java/hex/grid/Grid.java
Comment (Pscheidl) [inline h2o-core/src/main/java/hex/grid/Grid.java:549]: - The `saveTo` operation can potentially fail on a subset of persisted frames. There are runtimeExceptions thrown, which would cause the underlying Job to fail.
- `name` unused
- It is unclear whether this method blocks until all the frames are exported or not. This brings up yet another question - should be block on this operation (`exportFramesBinary`) and verify the output before moving on to next round ?

A rough proposal that both blocks and checks for errors (but only logs the errors, does not act upon them in any way):
```suggestion
    final List<Job<Frame>> crashedFrames = frames.values().stream()
            .map((f) -> new FramePersist(f).saveTo(exportDir, true))
            .filter(job -> {
              job.get(); // Wait for the persist operation ot complete.
              return job.isCrashed();
            })
            .collect(Collectors.toList());
    if(!crashedFrames.isEmpty()){
      crashedFrames.forEach(frame -> {
        Log.err(String.format("Unable to export binary frame '%s'", frame.get()._key.toString()));
      });
    }
```

File: h2o-core/src/main/java/hex/schemas/GridSearchSchema.java
Comment (Pscheidl) [inline h2o-core/src/main/java/hex/schemas/GridSearchSchema.java:56]: Thank you.

**Review by Pscheidl — COMMENTED** [review]
A couple of thoughts/questions.

File: h2o-core/src/main/java/water/fvec/persist/PersistUtils.java
Comment (Pscheidl) [reply h2o-core/src/main/java/water/fvec/persist/PersistUtils.java:34]: Yes, it's only a matter of preference - therefore a suggestion. I found it quite nice to declare exceptions thrown, even if Java doesn't force us to do so. It's your call, no big deal. It might as well stay as-is.

File: h2o-r/h2o-package/R/export.R
Comment (Pscheidl) [inline h2o-r/h2o-package/R/export.R:340]: Usually in R API, first the type expected is described. Not always, but most of the times yes. Just noticed it.

**Review by Pscheidl — APPROVED** [review]
LGTM, thanks. One minor thing to think about.

File: h2o-core/src/main/java/hex/grid/Grid.java
Comment (mn-mikke) [inline h2o-core/src/main/java/hex/grid/Grid.java:545]: just a question: What file systems are supported by `FramePersist`? Are save and load operations distributed? Any who looks after cleaning `exportDir` when training is properly done?

**Review by mn-mikke — COMMENTED** [review]
How ofter is grid search checkpointed? is it configurable?

What about `initial_weights` and `initial_biases` on `DeepLearningModel`? Do we need to persist them as well? + What `pretrained_autoencoder`?

File: h2o-algos/src/main/java/hex/gam/GAMModel.java
Comment (honzasterba) [reply h2o-algos/src/main/java/hex/gam/GAMModel.java:293]: interesting idea, AFAIK reflection does not contain information about generics, we could find all Key typed params but will not be able to find those that are Key<Frame>

File: h2o-core/src/main/java/hex/grid/Grid.java
Comment (honzasterba) [reply h2o-core/src/main/java/hex/grid/Grid.java:545]: all that are supported by rest of H2O for a distributed cluster a distributed FS should be used
save and load are distributed, each node writes/reads only local chunks, hence thre requirement for same cluster size on load
cleaning of exportDir is left up to the user, this is the same behaviour with normal export_checkpoints_dir option

File: h2o-algos/src/main/java/hex/gam/GAMModel.java
Comment (Pscheidl) [reply h2o-algos/src/main/java/hex/gam/GAMModel.java:293]: Or force a method on an algo.

Comment (sebhrusen) [reply h2o-algos/src/main/java/hex/gam/GAMModel.java:293]: @honzasterba see 
https://docs.oracle.com/en/java/javase/15/docs/api/java.base/java/lang/reflect/Field.html#getGenericType()
https://docs.oracle.com/en/java/javase/15/docs/api/java.base/java/lang/reflect/Method.html#getGenericParameterTypes()
https://docs.oracle.com/en/java/javase/15/docs/api/java.base/java/lang/reflect/ParameterizedType.html

all this is available since 1.5 and H2O-3 is actually using those in `ReflectionUtils`

Comment (honzasterba) [reply h2o-algos/src/main/java/hex/gam/GAMModel.java:293]: interesting, but this still does not help for the case of DeepLearningModel.initial_weights which is Key[], we could just take all Key and Key[] params and do DKV.get() and look at the object type and decide based on that

Comment (sebhrusen) [reply h2o-algos/src/main/java/hex/gam/GAMModel.java:293]: agree, this works only for code, there's still type erasure for instances of course, so you can't obtain any magic information from objects stored in DKV

File: h2o-core/src/main/java/hex/grid/Grid.java
Comment (mn-mikke) [reply h2o-core/src/main/java/hex/grid/Grid.java:545]: Thanks for your explanation!

File: h2o-algos/src/main/java/hex/gam/GAMModel.java
Comment (sebhrusen) [reply h2o-algos/src/main/java/hex/gam/GAMModel.java:293]: after reflection (pun intended), for cases like `Key[]` we could consider that they should be declared as `Key<Frame>[]` arrays, and then reflection would work...

Anyway, as much as I also like @mn-mikke suggestion, it may require too many changes, idk..

File: h2o-core/src/main/java/hex/grid/Grid.java
Comment (michalkurka) [inline h2o-core/src/main/java/hex/grid/Grid.java:539]: IMHO this code doesn't belong there, grid search shouldn't be itself responsible for its own recovery

instead, it should just say what objects or Keys are needed for resuming and propagate it to the layer above

we eventually want any process to be recoverable, I would expect GS to have a Recoverable interface that would include a method `Set<Key> getDependentObjects` - the layer avobe should be responsible for saving and re-loading those object upon restart

File: h2o-core/src/main/java/hex/grid/Grid.java
Comment (michalkurka) [inline h2o-core/src/main/java/hex/grid/Grid.java:540]: Should we instead use ModelParameterSchema class to get fields (and possibly their metadata)?

File: h2o-core/src/main/java/hex/grid/Grid.java
Comment (michalkurka) [inline h2o-core/src/main/java/hex/grid/Grid.java:587]: you keep exporting even after it already crashed because you are throwing an exception and skipping rest of the jobs - this behavior is definitely not ideal

simple sequential write would probably make more sense here

File: h2o-core/src/main/java/hex/grid/Grid.java
Comment (michalkurka) [inline h2o-core/src/main/java/hex/grid/Grid.java:570]: What would this be?

Comment (honzasterba) [reply h2o-core/src/main/java/hex/grid/Grid.java:570]: pretrained_autoencoder in DeepLearning for example

Comment (michalkurka) [reply h2o-core/src/main/java/hex/grid/Grid.java:570]: in this case, it would only write DeepLearningModel but not actually the weights and biases (that are in DKV), am I right?

File: h2o-core/src/main/java/hex/grid/Grid.java
Comment (honzasterba) [reply h2o-core/src/main/java/hex/grid/Grid.java]: fixed

File: h2o-core/src/main/java/hex/grid/Grid.java
Comment (honzasterba) [reply h2o-core/src/main/java/hex/grid/Grid.java]: refactored, please look it over

**Review by Pscheidl — APPROVED** [review]
Thanks.

File: h2o-core/src/main/java/hex/grid/GridSearch.java
Comment (michalkurka) [inline h2o-core/src/main/java/hex/grid/GridSearch.java:462]: @honzasterba IMHO this is mixing two concepts together

Parameter `export_checkpoints_dir` is meant for the _user_ to export models - this can have a different motivation than just recovery.

Fault-tolerance is driven by the software, not the user. I think there should be 2 locations - one for user checkpointed objects (that one is preserved after modeling ends) and one for fault-tolerance objects (that was is cleaned up automatically).

File: h2o-core/src/main/java/hex/grid/GridSearch.java
Comment (michalkurka) [inline h2o-core/src/main/java/hex/grid/GridSearch.java:91]: instead of passing this parameter, I think it makes more logical sense to pass the recovery object - where the grid can store recovery information; if this parameter is not passed - nothing happens

File: h2o-py/h2o/grid/grid_search.py
Comment (Pscheidl) [inline h2o-py/h2o/grid/grid_search.py:79]: This API requires the user to specify the dir. What if H2O is managed ? I believe in managed runs, like on Kubernetes, an automatic regime would be useful. How are we going to solve this ? By introducing another parameter in future, e.g. `autorecovery` ? 

Could be solved by enums in Py (not sure how to solve this in R, but S4 classes seems to offer a solution) and introduce two modes for a single `recovery` parameter. Pseudocode:

enum Recovery{
Directory(Path directoryToSaveTo),
Automated
}

The automated mode could simply dump it to a well known location based on system, e.g. /var/h2o on linux.

File: h2o-core/src/main/java/hex/faulttolerance/Recovery.java
Comment (honzasterba) [reply h2o-core/src/main/java/hex/faulttolerance/Recovery.java:23]: I think that is out of scope for this PR, everywhere we use strings to pass URIs so that would have to be a bigger change

File: h2o-core/src/main/java/water/fvec/persist/PersistUtils.java
Comment (honzasterba) [reply h2o-core/src/main/java/water/fvec/persist/PersistUtils.java:37]: believe me I tried adding it! but autobuffer does so many things that I gave up

File: h2o-core/src/main/java/hex/faulttolerance/Recoverable.java
Comment (honzasterba) [reply h2o-core/src/main/java/hex/faulttolerance/Recoverable.java:19]: this is not part of public API so, for now I think we are good and since another target is autoML I think one file should be enough, we can alway change this if it is not sufficient

File: h2o-py/h2o/grid/grid_search.py
Comment (honzasterba) [reply h2o-py/h2o/grid/grid_search.py:79]: This API is intended for purely manual use case - see the tests that use it, basically the user is responsible for starting a new cluster and loading the grid and continuing training.
For K8s and other auto-recoverable envs I would imagine this would be done somewhat automatically by the back-end. So the user would only say the cluster should auto-recover and all this magic will happen behind the scenes.

Comment (Pscheidl) [reply h2o-py/h2o/grid/grid_search.py:79]: I see.  Seems fine to me.

**Review by Pscheidl — APPROVED** [review]

File: h2o-py/tests/testdir_algos/grid/pyunit_grid_fault_tolerance_resume.py
Comment (michalkurka) [inline h2o-py/tests/testdir_algos/grid/pyunit_grid_fault_tolerance_resume.py:319]: is this commented out intentionally?

File: h2o-py/tests/testdir_algos/grid/pyunit_grid_manual_export_resume_new_hyperspace.py
Comment (michalkurka) [inline h2o-py/tests/testdir_algos/grid/pyunit_grid_manual_export_resume_new_hyperspace.py:23]: Why this change? It is better to output file to results directory, then they are preserved if anything fails.

File: h2o-r/h2o-package/R/grid.R
Comment (michalkurka) [inline h2o-r/h2o-package/R/grid.R:43]: this seems to be called h2o.loadGrid

File: h2o-core/src/main/java/water/api/schemas3/GridExportV3.java
Comment (michalkurka) [inline h2o-core/src/main/java/water/api/schemas3/GridExportV3.java:16]: I think we need to be more clear what is actually saved. Users can easily confuse this for eg. CV holdout predictions or other frames that are created as a _result_ of model training.

File: h2o-py/tests/testdir_algos/grid/pyunit_grid_manual_export_resume_new_hyperspace.py
Comment (honzasterba) [reply h2o-py/tests/testdir_algos/grid/pyunit_grid_manual_export_resume_new_hyperspace.py:23]: I thought only test results are to be written to the results dir, tmp data should be in tempfiles, same for junit

File: h2o-r/h2o-package/R/grid.R
Comment (honzasterba) [reply h2o-r/h2o-package/R/grid.R]: fixed

File: h2o-core/src/main/java/water/api/schemas3/GridExportV3.java
Comment (honzasterba) [reply h2o-core/src/main/java/water/api/schemas3/GridExportV3.java]: good call, I tried to improve the naming and docs

**Review by hannah-tillman — APPROVED** [review]

File: h2o-py/tests/testdir_algos/grid/pyunit_grid_manual_export_resume_new_hyperspace.py
Comment (michalkurka) [reply h2o-py/tests/testdir_algos/grid/pyunit_grid_manual_export_resume_new_hyperspace.py:23]: That depends on what is tmp data to you but this use-case is imho similar to eg. checkpointing and those tests use pyunit_utils.locate("results"), example https://github.com/h2oai/h2o-3/blob/master/h2o-py/tests/testdir_algos/xgboost/pyunit_checkpoint_xgboost.py#L30

For data exchange for tests we follow the pyunit_utils.locate("results") method because it has a clear lifecycle.

The temp file will be gone with the docker exiting on Jenkins which could be a bad thing for reproducing a bug.

**Review by michalkurka — APPROVED** [review]
Thank you @honzasterba

**Pull request merged**
